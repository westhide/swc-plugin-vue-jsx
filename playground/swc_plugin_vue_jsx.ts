import { type Plugin, createFilter } from "vite";
import { transformSync } from "@swc/core";

export default function () {
  const filter = createFilter(/\.[jt]sx$/);

  return {
    name: "swc-plugin-vue-jsx",
    enforce: "pre",

    async transform(src, id) {
      if (filter(id)) {
        return transformSync(src, {
          isModule: true,
          jsc: {
            target: "es2022",
            parser: {
              syntax: "typescript",
              tsx: true,
            },
            experimental: {
              plugins: [["@westhide/swc-plugin-vue-jsx", {}]],
            },
          },
        });
      } else {
        return null;
      }
    },
  } as Plugin;
}
