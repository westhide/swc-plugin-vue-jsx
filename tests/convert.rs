use insta::assert_snapshot;
use swc_core::{
    common::{chain, Mark},
    ecma::{
        parser::{Syntax, TsConfig},
        transforms::{
            base::{fixer::fixer, hygiene::hygiene},
            testing::Tester,
        },
        visit::as_folder,
    },
    plugin::proxies::PluginCommentsProxy,
};
use swc_plugin_vue_jsx::{PluginOptions, VueJSX};

const TSX_SYNTAX: Syntax = Syntax::Typescript(TsConfig {
    tsx: true,
    decorators: false,
    dts: false,
    no_early_errors: false,
});

macro_rules! test {
    ($name:ident, $src:literal, $opts:expr) => {
        #[test]
        #[allow(non_snake_case)]
        fn $name() {
            Tester::run(|tester| {
                let module = tester.apply_transform(
                    chain!(
                        as_folder(VueJSX::<PluginCommentsProxy>::new(
                            $opts,
                            None,
                            Mark::new(),
                        )),
                        hygiene(),
                        fixer(Some(&tester.comments.clone()))
                    ),
                    "test.tsx",
                    TSX_SYNTAX,
                    $src,
                )?;

                let code = tester.print(&module, &tester.comments.clone());

                assert_snapshot!(stringify!($name), code);

                Ok(())
            });
        }
    };

    ($name:ident, $src:literal) => {
        test!($name, $src, PluginOptions::from("{}"));
    };

    ({ $($name:ident : $src:literal),+ $(,)? }) => {
        $(test!($name,$src);)+
    };
}

test!({
    HelloWorld: r#"const app = () => <h1>Hello World</h1>"#
});
