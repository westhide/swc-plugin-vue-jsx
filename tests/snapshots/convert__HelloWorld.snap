---
source: tests/convert.rs
expression: code
---
import { createTextVNode, createElementVNode } from "vue";
const _hoisted_ = createTextVNode("Hello World"), _hoisted_1 = createElementVNode("h1", null, [
    _hoisted_
], -1);
const app = ()=>_hoisted_1;

